src/libertem/io/dataset/dask.py

```diff
@@ -0,0 +1,483 @@
+                array = da.transpose(array, axes=sort_order)
+                log.info('Re-ordered nav_dimensions to improve partitioning, '
+                         'create the dataset with preserve_dimensions=True '
+                         'to suppress this behaviour. '
+                         f'Original shape: {original_shape} with '
+                         f'n_blocks: {original_n_chunks}. '
+                         f'New shape: {array.shape} with '
+                         f'n_blocks: {[len(c) for c in array.chunks]}.')
+        # Handle chunked nav_dimensions
+        # We can allow nav_dimensions to be fully chunked (one chunk per element)
+        # up-to-but-not-including the first non-fully chunked dimension. After this point
+        # we must merge/rechunk all subsequent nav dimensions to ensure continuity
+        # of frame indexes in a flattened nav dimension. This should be removed
+        # when if we allow non-contiguous flat_idx Partitions
+        nav_rechunk_dict = {}
+        for dim_idx, dim_chunking in enumerate(array.chunks[:-sig_dims]):
+            if set(dim_chunking) == {1}:
+                continue
+            else:
+                merge_dimensions = [*range(dim_idx + 1, n_dimension - sig_dims)]
+                for merge_i in merge_dimensions:
+                    if len(array.chunks[merge_i]) > 1:
+                        nav_rechunk_dict[merge_i] = -1
+        if nav_rechunk_dict:
+            original_n_chunks = [len(c) for c in array.chunks]
+            array = array.rechunk(nav_rechunk_dict)
+            log.info('Merging nav dimension chunks according to scheme '
+                     f'{nav_rechunk_dict} as we cannot maintain continuity '
+                     'of frame indexing in the flattened navigation dimension. '
+                     f'Original n_blocks: {original_n_chunks}. '
+                     f'New n_blocks: {[len(c) for c in array.chunks]}.')
+        # Merge remaining chunks maintaining C-ordering until we reach a target chunk sizes
+        # or a minmum number of partitions corresponding to the number of workers
+        new_chunking, min_size, max_size = merge_until_target(array, self._min_size,
+                                                              self._min_npart)
+        if new_chunking != array.chunks:
+            original_n_chunks = [len(c) for c in array.chunks]
+            chunksizes = get_chunksizes(array)
+            orig_min, orig_max = chunksizes.min(), chunksizes.max()
+            array = array.rechunk(new_chunking)
+            log.info('Applying re-chunking to increase minimum partition size. '
+                     f'n_blocks: {original_n_chunks} => {[len(c) for c in array.chunks]}. '
+                     f'Min chunk size {orig_min / 1e6:.1f} => {min_size / 1e6:.1f} MiB , '
+                     f'Max chunk size {orig_max / 1e6:.1f} => {max_size / 1e6:.1f} MiB.')
+        return array
```

Comment:
We have similar code used in https://github.com/LiberTEM/LiberTEM/blob/d294d5d7b06dcb3ea6954c804b6eef53dc281fa5/src/libertem/io/dataset/hdf5.py#L390-L409

Maybe some of these functions can be re-used here? Possibly one can move https://github.com/LiberTEM/LiberTEM/blob/d294d5d7b06dcb3ea6954c804b6eef53dc281fa5/src/libertem/io/dataset/hdf5.py#L106-L133 to `libertem.io.utils`?

Reply:
I'm looking into this. I believe the biggest difference I've seen so far is that in the current code I handle arrays with arbitrary chunking along an axis (not just uniform chunks with a remainder), but I need to see where we can make the code common.

Reply:
Good point about the variable chunk size! Also, chunks in HDF5 can be quite small compared to Dask array chunks so it would not be advisable to iterate over all of them.

Reply:
It seems like there is no support for non-uniform chunksizes along an axis in HDF5 except for the remainder chunk at the end.

When exporting using `da.to_hdf5` Dask or h5py seems to re-chunk each axis so that the normal chunksize is the same as the first chunksize, then lumps everything else into a remainder:

```python
dar = da.ones((59, 32, 32, 32), chunks=((12, 45, 2), -1, -1, -1))
da.to_hdf5('file.hdf5', '/data', dar)
print(h5py.File('file.hdf5')['data'].chunks)
>>> (12, 32, 32, 32)
```

Reply:
Ok, sounds like the best option is to keep separate functions for HDF5 and Dask